pkg/auth/dependency/webapp/validate_provider.go

```diff
@@ -0,0 +1,13 @@
+package webapp
+
+import (
+	"net/url"
+)
+
+// ValidateProvider validates url.Values with JSON Schema.
+type ValidateProvider interface {
+	// Prevalidate fills in default values.
+	Prevalidate(form url.Values)
```

Comment:
Maybe name it `SetDefaults` instead? `Prevalidate` is rather confusing.

Reply:
But it also removes empty values.

Reply:
How about `PrepareValues`? It prepares the form values for validation & using in provider.